zebra-chain/src/primitives/redpallas/tests/prop.rs

```diff
@@ -0,0 +1,155 @@
+       nonce commitment is fed into the hash, so it has to have torsion at signing
+       time.
+    /// Change the case to have a torsion component in the signature's `r` value.
+    AddTorsion,
+    */
+    /* XXX implement this -- needs custom handling of field arithmetic.
+    /// Change the signature's `s` scalar to be unreduced (mod L), invalidating the signature.
+    UnreducedScalar,
+    */
+}
+
+impl<T: SigType> SignatureCase<T> {
+    fn new<R: RngCore + CryptoRng>(mut rng: R, msg: Vec<u8>) -> Self {
+        let sk = SigningKey::new(&mut rng);
+        let sig = sk.sign(&mut rng, &msg);
+        let pk_bytes = VerificationKey::from(&sk).into();
+        Self {
+            msg,
+            sig,
+            pk_bytes,
+            is_valid: true,
+        }
+    }
+
+    // Check that signature verification succeeds or fails, as expected.
+    fn check(&self) -> bool {
+        // The signature data is stored in (refined) byte types, but do a round trip
+        // conversion to raw bytes to exercise those code paths.
+        let sig = {
+            let bytes: [u8; 64] = self.sig.into();
+            Signature::<T>::from(bytes)
+        };
+        let pk_bytes = {
+            let bytes: [u8; 32] = self.pk_bytes.into();
+            VerificationKeyBytes::<T>::from(bytes)
+        };
+
+        // Check that signature validation has the expected result.
+        self.is_valid
+            == VerificationKey::try_from(pk_bytes)
+                .and_then(|pk| pk.verify(&self.msg, &sig))
+                .is_ok()
+    }
+
+    fn apply_tweak(&mut self, tweak: &Tweak) {
+        match tweak {
+            Tweak::None => {}
+            Tweak::ChangeMessage => {
+                // Changing the message makes the signature invalid.
+                self.msg.push(90);
+                self.is_valid = false;
+            }
+            Tweak::ChangePubkey => {
+                // Changing the public key makes the signature invalid.
+                let mut bytes: [u8; 32] = self.pk_bytes.into();
+                let j = (bytes[2] & 31) as usize;
+                bytes[2] ^= 0x23;
+                bytes[2] |= 0x99;
+                bytes[j] ^= bytes[2];
+                self.pk_bytes = bytes.into();
+                self.is_valid = false;
+            }
+        }
+    }
+}
+
+fn tweak_strategy() -> impl Strategy<Value = Tweak> {
+    prop_oneof![
+        10 => Just(Tweak::None),
+        1 => Just(Tweak::ChangeMessage),
+        1 => Just(Tweak::ChangePubkey),
+    ]
+}
+
+proptest! {
+    #[test]
+    fn tweak_signature(
+        tweaks in prop::collection::vec(tweak_strategy(), (0,5)),
+        rng_seed in any::<u64>(),
+    ) {
+        // Use a deterministic RNG so that test failures can be reproduced.
+        // Seeding with 64 bits of entropy is INSECURE and this code should
+        // not be copied outside of this test!
+        let mut rng = ChaChaRng::seed_from_u64(rng_seed);
```

Comment:
Sometimes it's safer just to use the correct code :-)

Reply:
Fixed! 7b3ad86a

Reply:
Yep you know how people love to copy code (and not read or copy comments) 🙃